File: solutions/lua/transpose/1/transpose.lua
-- StringToMatrix converts the input string to a matrix (each row in the matrix is delimited
--  by a newline in the input).
local function stringToMatrix(s)

    local matrix = {}

    for line in s:gmatch('[^\n]*') do
        local row = {}
        table.insert(matrix, row)
        for cell in line:gmatch('.') do
            table.insert(row, cell)
        end
    end

    return matrix
end

-- MaxRowSize returns the size of the longuest row in the matrix.
local function maxRowSize(matrix)

    local max = 0
    for _, row in ipairs(matrix) do
        if #row > max then
            max = #row
        end
    end

    return max
end

-- Transpose transposes the matrix and pad to the left but not to the right.
local function transpose(matrix)

    local transposed = {}
    local ncols = maxRowSize(matrix)

    -- initialize the transposed matrix
    for r = 1, ncols do
        transposed[r] = {}
    end

    -- to pad to the left, one needs to start filling the transpose matrix column first
    -- and start with the right-most column.
    -- Once a column has been filled with N elements, all left columns must contain at least N
    -- elements (left padding)/
    max = 0
    for c = #matrix, 1, -1 do
        max = math.max(max, #matrix[c])
        for r = 1, max do
            transposed[r][c] = matrix[c][r] or ' '
        end
    end

    return transposed
end

-- MatrixToString converts a matrix back to a string with each row being separated with
-- a newline.
local function matrixToString(matrix)

    local s = {}

    for _, row in ipairs(matrix) do
        --local line = table.concat(row)
        --line = trimRight(line)
        table.insert(s, table.concat(row))
    end

    return table.concat(s, '\n')
end

return function(s)

    return matrixToString(transpose(stringToMatrix(s)))

end

File: solutions/javascript/transpose/1/transpose.js

export function transpose(strings) {
  
  if (strings.length == 0) return [];
  
  // To transpose a matrix, it needs to be rectangular (i.e. all rows must be of
  // the same length).
  // If all the lines in the original matrix are not of the same length, then we will
  // right pad the shorter lines with spaces.
  // There is an exception: If the set of last original lines are shorter, don't pad them.
  const maxLength = strings.reduce((max, string) => Math.max(string.length, max), 0);

  let result = [];
  for (let c = 0; c < maxLength; c++) {
    const rowCharacters = [];
    for (let i = 0; i < strings.length; i++) {
      rowCharacters.push(strings[i].charAt(c) || ' ');
    }
    let row = rowCharacters.join('');
    // clean up trailing blanks on the last row
    if (c == maxLength-1) row = row.trimRight();
    result.push(row);
  }
  return result;
};
